File: backend/app/api/v1/endpoints/templates.py
from __future__ import annotations

import uuid
from typing import Any

from fastapi import APIRouter, Depends

from app.api.deps import get_current_active_user, get_unit_of_work
from app.application.dto.templates import (
    CreateTemplateRequest,
    TemplateListResponse,
    TemplateResponse,
    UpdateTemplateRequest,
)
from app.application.services.template_service import TemplateService
from app.infrastructure.persistence.models.user import User

router = APIRouter(prefix="/api/v1/templates", tags=["Templates"])


def get_template_service(uow=Depends(get_unit_of_work)) -> TemplateService:
    return TemplateService(uow)


@router.get("", response_model=TemplateListResponse, summary="List templates")
async def list_templates(
    current_user: User = Depends(get_current_active_user),
    service: TemplateService = Depends(get_template_service),
) -> Any:
    return await service.get_templates()


@router.get("/code/{code}", response_model=TemplateResponse, summary="Get template by code (e.g. K9)")
async def get_template_by_code(
    code: str,
    current_user: User = Depends(get_current_active_user),
    service: TemplateService = Depends(get_template_service),
) -> Any:
    return await service.get_template_by_code(code)


@router.get("/{id}", response_model=TemplateResponse, summary="Get template by ID")
async def get_template(
    id: uuid.UUID,
    current_user: User = Depends(get_current_active_user),
    service: TemplateService = Depends(get_template_service),
) -> Any:
    return await service.get_template(id)


@router.post("", response_model=TemplateResponse, summary="Create a new template")
async def create_template(
    data: CreateTemplateRequest,
    current_user: User = Depends(get_current_active_user),
    service: TemplateService = Depends(get_template_service),
) -> Any:
    return await service.create_template(data)


@router.put("/{id}", response_model=TemplateResponse, summary="Update a template")
async def update_template(
    id: uuid.UUID,
    data: UpdateTemplateRequest,
    current_user: User = Depends(get_current_active_user),
    service: TemplateService = Depends(get_template_service),
) -> Any:
    return await service.update_template(id, data)


@router.post("/{id}/duplicate", response_model=TemplateResponse, summary="Duplicate a template")
async def duplicate_template(
    id: uuid.UUID,
    current_user: User = Depends(get_current_active_user),
    service: TemplateService = Depends(get_template_service),
) -> Any:
    return await service.duplicate_template(id)


@router.post("/{id}/publish", response_model=TemplateResponse, summary="Publish a template")
async def publish_template(
    id: uuid.UUID,
    current_user: User = Depends(get_current_active_user),
    service: TemplateService = Depends(get_template_service),
) -> Any:
    return await service.publish_template(id)


@router.post("/{id}/archive", response_model=TemplateResponse, summary="Archive a template")
async def archive_template(
    id: uuid.UUID,
    current_user: User = Depends(get_current_active_user),
    service: TemplateService = Depends(get_template_service),
) -> Any:
    return await service.archive_template(id)


@router.delete("/{id}", summary="Delete a template")
async def delete_template(
    id: uuid.UUID,
    current_user: User = Depends(get_current_active_user),
    service: TemplateService = Depends(get_template_service),
) -> dict[str, bool]:
    return {"success": await service.delete_template(id)}


@router.post("/import", response_model=TemplateResponse, summary="Import template JSON")
async def import_template_json(
    raw_json: dict[str, Any],
    current_user: User = Depends(get_current_active_user),
    service: TemplateService = Depends(get_template_service),
) -> Any:
    return await service.import_template_json(raw_json)


@router.get("/{id}/export", summary="Export template JSON")
async def export_template_json(
    id: uuid.UUID,
    current_user: User = Depends(get_current_active_user),
    service: TemplateService = Depends(get_template_service),
) -> dict[str, Any]:
    tmpl = await service.get_template(id)
    return tmpl.schema_json

File: frontend/src/api/templates.ts
import { api } from './client';
import { API_BASE_URL } from '@/lib/constants';
import { CMFTemplate } from '@/types/template';

// NOTE: All paths here are relative to API_BASE_URL (/api/v1).
// Do NOT include /api/v1 prefix — the api client adds it automatically.

const TOKEN_KEY = 'cmf_access_token';

/**
 * Posts a JSON body to the backend WITHOUT snake_case transformation.
 * Required when the payload contains a `schema_json` domain blob whose inner
 * keys must be preserved exactly as authored (e.g. `internalName`, `helpText`,
 * `defaultValue`) — the shared `api.post()` converts those to snake_case.
 */
async function rawJsonPost<T>(url: string, body: unknown): Promise<T> {
  const token = localStorage.getItem(TOKEN_KEY) || '';
  const response = await fetch(`${API_BASE_URL}${url}`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      ...(token ? { Authorization: `Bearer ${token}` } : {}),
    },
    body: JSON.stringify(body),
  });

  if (!response.ok) {
    let message = `Request failed with status ${response.status}`;
    try {
      const errBody = await response.json();
      if (errBody?.detail) {
        message =
          typeof errBody.detail === 'string'
            ? errBody.detail
            : errBody.detail?.message || message;
      } else if (errBody?.message) {
        message = errBody.message;
      } else if (errBody?.error?.message) {
        message = errBody.error.message;
      }
    } catch {
      // ignore parse errors, keep default message
    }
    throw new Error(message);
  }

  return response.json() as Promise<T>;
}

export const templatesApi = {
  getTemplates: async (): Promise<{ items: CMFTemplate[]; total: number }> => {
    return api.get('/templates');
  },

  getTemplateById: async (id: string): Promise<CMFTemplate> => {
    return api.get(`/templates/${id}`);
  },

  getTemplateByCode: async (code: string): Promise<CMFTemplate> => {
    return api.get(`/templates/code/${code}`);
  },

  createTemplate: async (data: {
    code: string;
    name: string;
    description?: string;
    version?: string;
    status?: string;
    schema_json: any;
  }): Promise<CMFTemplate> => {
    return api.post('/templates', data);
  },

  /**
   * Creates a Template (Project Structure) sending the full body verbatim —
   * preserves camelCase keys inside `schema_json` (internalName, helpText, etc.)
   */
  createStructureTemplate: async (data: {
    code: string;
    name: string;
    description?: string;
    version?: string;
    status?: string;
    schema_json: any;
  }): Promise<CMFTemplate> => {
    return rawJsonPost<CMFTemplate>('/templates', data);
  },

  updateTemplate: async (
    id: string,
    data: {
      name?: string;
      description?: string;
      status?: string;
      schema_json?: any;
      change_log?: string;
    }
  ): Promise<CMFTemplate> => {
    return api.put(`/templates/${id}`, data);
  },

  duplicateTemplate: async (id: string): Promise<CMFTemplate> => {
    return api.post(`/templates/${id}/duplicate`);
  },

  publishTemplate: async (id: string): Promise<CMFTemplate> => {
    return api.post(`/templates/${id}/publish`);
  },

  archiveTemplate: async (id: string): Promise<CMFTemplate> => {
    return api.post(`/templates/${id}/archive`);
  },

  deleteTemplate: async (id: string): Promise<{ success: boolean }> => {
    return api.delete(`/templates/${id}`);
  },

  importTemplateJson: async (rawJson: any): Promise<CMFTemplate> => {
    return api.post('/templates/import', rawJson);
  },

  /**
   * Imports a Template (Project Structure) from raw JSON verbatim.
   */
  importTemplateRawJson: async (rawJson: any): Promise<CMFTemplate> => {
    return rawJsonPost<CMFTemplate>('/templates/import', rawJson);
  },

  exportTemplateJson: async (id: string): Promise<any> => {
    return api.get(`/templates/${id}/export`);
  },
};
